### PR TITLE
Data Lake CPK

### PR DESCRIPTION
**Pull request by seanmcc-msft**

### CODE REVIEW

File: sdk/storage/Azure.Storage.Files.DataLake/src/autorest.md [hunk 1]
```diff
@@ -4,7 +4,7 @@ Run `dotnet build /t:GenerateCode` to generate code.
 
 ``` yaml
 input-file:
-    - https://raw.githubusercontent.com/Azure/azure-rest-api-specs/4a93ab078fba7f087116283c8ed169f9b8e30397/specification/storage/data-plane/Microsoft.StorageDataLake/stable/2020-06-12/DataLakeStorage.json
+    - C:\Users\seanmcc\git\azure-rest-api-specs\specification\storage\data-plane\Microsoft.StorageDataLake\preview\2020-10-02\DataLakeStorage.json
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.DataLake/src/autorest.md:7]: TODO undo this.

File: sdk/storage/Azure.Storage.Files.DataLake/api/Azure.Storage.Files.DataLake.netstandard2.0.cs [hunk 2]
```diff
@@ -418,6 +422,23 @@ public enum CopyStatus
         Aborted = 2,
         Failed = 3,
     }
+    [System.Runtime.InteropServices.StructLayoutAttribute(System.Runtime.InteropServices.LayoutKind.Sequential)]
+    public readonly partial struct CustomerProvidedKey : System.IEquatable<Azure.Storage.Files.DataLake.Models.CustomerProvidedKey>
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.DataLake/api/Azure.Storage.Files.DataLake.netstandard2.0.cs:426]: Add DataLake prefix.

File: sdk/storage/Azure.Storage.Files.DataLake/api/Azure.Storage.Files.DataLake.netstandard2.0.cs [hunk 3]
```diff
@@ -675,6 +696,10 @@ public partial class DataLakeStaticWebsite
         public string ErrorDocument404Path { get { throw null; } set { } }
         public string IndexDocument { get { throw null; } set { } }
     }
+    public enum EncryptionAlgorithmType
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.DataLake/api/Azure.Storage.Files.DataLake.netstandard2.0.cs:699]: Add DataLake prefix.

File: sdk/storage/Azure.Storage.Files.DataLake/api/Azure.Storage.Files.DataLake.netstandard2.0.cs (hunk 2 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.DataLake/api/Azure.Storage.Files.DataLake.netstandard2.0.cs:426]: Fixed.

File: sdk/storage/Azure.Storage.Files.DataLake/api/Azure.Storage.Files.DataLake.netstandard2.0.cs (hunk 3 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.DataLake/api/Azure.Storage.Files.DataLake.netstandard2.0.cs:699]: Fixed.

File: sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeErrors.cs [hunk 4]
```diff
@@ -59,5 +59,13 @@ public static DataLakeAclChangeFailedException ChangeAclFailed(Exception excepti
                 $"{nameof(DataLakeAclChangeFailedException.ContinuationToken)}={continuationToken} after addressing the error.",
                 exception,
                 continuationToken);
+
+        internal static void VerifyHttpsCustomerProvidedKey(Uri uri, DataLakeCustomerProvidedKey? customerProvidedKey)
```
Comment (jaschrep-msft) [inline sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeErrors.cs:63]: Do we normally do verification in the errors class? I feel like we usually just assemble them here when it has already been determined we need to throw.

**Review by jaschrep-msft — APPROVED** [review]
lgtm but a pass from someone more familiar with cpk as a whole would be good as well

File: sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeErrors.cs (hunk 4 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeErrors.cs:63]: Good question, in this case I pattern matched [BlobErrors](https://github.com/Azure/azure-sdk-for-net/blob/9d2509baa0d972971518ad43483f3a35937a216a/sdk/storage/Azure.Storage.Blobs/src/BlobErrors.cs#L26).

**Review by amnguye — APPROVED** [review]

**Pull request merged**